### ISSUE TITLE
RFC: plugin architecture

### ISSUE THREAD

**Issue by nmiyake** [palantir/godel#234]
This RFC is for the ability to add plugins to gödel. The proposal is to implement a plugin architecture for gödel in 2 phases: the first phase will add support for adding new tasks as plugins. The second phase will convert most or all of the existing tasks (`build`, `dist` etc.) to use the plugin architecture as well.

Motivation
========
Currently, all of the tasks provided by gödel must be part of the gödel binary itself. This works fine for core functionality that is meant to be shared by all users of gödel. However, it does not allow for gödel tasks to be customized. For example, consider the case where an organization wants to add a task that uses a proprietary library or provides custom functionality for that organization. Right now, there is no great way to do so -- the best approximation would be to structure the task as a `go generate` task and to hook it up there, but this is not very satisfying.

It would also be nice to have the ability to configure tasks as well. For example, right now the set of static checks run in `check` is hard-coded (adding or removing a check requires modifying gödel itself). Similarly, adding a distribution type or publish destination type requires modifying core code, which can be hard (or impossible if the formats or services are proprietary).

Proposal: task plugins and assets
=========================
I propose the following framework as a solution to this issue:

* A plugin is simply a program that can be invoked/run
  * Running a plugin with the `__godelPluginInfo` command prints the plugin metadata to `stdout`
* Allow plugins to be specified as part of configuration
* Plugin configuration can themselves specify assets that can customize plugin behavior
* When gödel is run, it retrieves the plugins and assets defined in configuration (the config specifies a URL/resolver pattern)
* Plugins provide information on the command(s) they provide
* gödel performs basic verification to ensure that there are no conflicts between plugins
* If the arguments provided to gödel match the plugin path, gödel invokes the plugin executable and provides all arguments after it (flag and non-flag arguments) to the plugin
  * It also provides the `--json` flag, whose value is some JSON configuration (includes an `exclude` block and other common configuration)
  * It also provides the `--config` flag and provides the path to the configuration file for the plugin
  * If the plugin configuration specified assets, a path to the assets directory is also provided
* At this point, the plugin program performs its normal execution

Implementation
===========

Read plugin configuration and download
----------------------------------------
The first component of implementation is defining a way to declare and retrieve plugins. As a proposal, consider `godel/config/plugins.yml` with content such as:

```yaml
plugins:
  - id: com.org.custom.foo
    product: foo
    version: 1.0.0
    resolver: https://internal.org.com/provider/{{product}}-{{os-arch}}-{{version}}.tgz
```

The resolver/URL can be anything, but the downloaded result must be an executable (or an archive that contains only an executable) that can be run with the `__godelPluginInfo` command and returns the expected values.

When gödel is run, it reads the `plugins.yml` file and first verifies that all of the plugin names are unique -- if they are not, the program fails with a configuration error.

Then, gödel will check its local cache for plugins: for example, in the case above, it would check for `~/.godel/plugins/com.org.custom.foo-darwin-amd64-1.0.0`. If it doesn't exist, the plugin is downloaded from the resolver.

After the plugin is located, it is executed with the `__godelPluginInfo` argument. The output is the JSON representation of a `PluginInfo` object:

```go
type PluginInfo struct {
        SchemaVersion string
	ID string
	Version string
	ConfigFileName string
	Commands map[string][]string
}
```

The portions of particular interest are `ConfigFileName` and `Commands`. "Commands" provides a mapping from the gödel command(s) that the plugin provides to the set of commands that correspond to the plugin.

As an example, assume that the plugin above returns the following JSON:

```json
SchemaVersion: 1
ID: com.org.custom.foo
Version: 1.0.0
ConfigFileName: foo.yml
Commands:
  foo: [foo]
  bar: [other, bar]
```

The `ConfigFileName` indicates that the configuration file is named `foo.yml`. The `Commands` indicates that the plugin provides the `foo` and `bar` commands to gödel, and that those commands correspond to the `foo` and `other bar` commands for the plugin program, respectively.

If multiple plugins provide the same command, that is considered a configuration error (for example, another plugin cannot also provide `foo` as a command).

Plugin execution
-----------------
Once the plugin configuration is read, gödel hooks up the CLI commands appropriately. In this case, it will register a `cli.Command` for `foo` and `bar`. `foo` will invoke the plugin executable with the argument `foo` (followed by any other flags/arguments provided by the user), while `bar` will invoke the plugin with `other bar` (+ user-provided input).

Additionally, the plugin executables are always provided with 2 other flags: if the `ConfigFileName` property is non-empty, it passes `--config=[path-to-config]`, and it passes `--json=[json of godel common config]`. Currently, the "godel common config" is the `exclude:` map defined in `exclude.yml`, but this may be expanded to include more things later.

So, if `./godelw bar --my-flag arg1 arg2` is executed, this results in gödel executing: `<path-to-plugin> --json='{exclude:...}' --config=godel/config/foo.yml other bar --my-flag arg1 arg2`.

Transitioning existing tasks to plugin framework
===================================
The framework outlined above can be defined and implemented additively for new user-defined tasks. However, it is also flexible enough such that existing built-in gödel tasks should be able to use it. For example, `distgo` would provide `build`, `dist`, `docker`, `publish`, while `gunit` would provide `test`. This would allow the subproducts to become more isolated, and gödel itself shrinks its concern to being just a framework for downloading and orchestrating the component plugins (although the default configuration would likely include the set of plugins that provide the current core functionality).

Assets (plugins for plugins)
====================
One aspect not covered by the above is customizing the behavior of a plugin. For example, if `okgo` became a plugin, it would be nice to make the checks themselves pluggable. Similarly, it would be nice if custom packagers or uploaders could be defined for `distgo`.

The task provider interface for gödel is fairly well-defined -- we always want to call some command, and we will automatically supply the required flags. However, individual plugin implementations may have more specialized needs. It is hard to define a generic interface for this, so instead we will take the approach of allowing assets to be defined and make them available to the plugin. The plugin can then use those assets however they see fit.

For example, consider the following configuration:

```yaml
  - id: com.palantir.distgo
    product: distgo
    org: palantir
    group: godel
    version: 1.0.0
    resolver: <some-resolver>
    assets:
      id: com.org.custom.dmg
      product: dmg-packager
      version: 0.5.3
      resolver: <some-resolver>
```

When `distgo` is executed, gödel will first ensure that all the assets specified in the `assets` block are retrieved/available, place them in a specific directory (or symlinks to them), and then provide that directory to the plugin using the `--assets` flag.

The plugin is then free to define whatever contracts it wants for its assets. For example, it could use a framework like [`go-plugin`](https://github.com/hashicorp/go-plugin/) and require that its assets be programs that provide this functionality and have the plugin program load the asset directory as its own plugin directory.

Implementation Roadmap
===================
* Define YML format for plugins (without assets)
* Define JSON structure that is output by `__godelPluginInfo`
* Implement functionality for reading plugin configuration and fetching plugins
* Add framework for hooking up plugins as commands
--> The above should be sufficient for the functionality for adding new commands as plugins
* Define YML format for assets
* Implement functionality for fetching assets and making them available to plugins
--> The above should be sufficient for adding plugins with assets
* Convert core gödel commands to use plugin framework

Adding support for plugins and assets should be possible as purely additive work that does not break any backwards-compatibility. However, converting core gödel commands will likely require some breaks (although the transition should be made as smoothly as possible) and may require a major version bump.

**Comment by nmiyake** [comment]
@bmoylan @jlz27 @ryanmcnamara @jonsyu1 @hughyoh @gdearment for SA. I generally feel pretty good about the approach, but feedback welcome!

**Comment by nmiyake** [comment]
Another note: plugin metadata should also include whether or not a plugin task should be part of the `verify` task. If it is, it should also specify how things should be run for `--apply=true` versus `--apply=false`.

**Comment by nmiyake** [comment]
Implemented mostly as described above in 1.1.0.
